reana_workflow_engine_serial/publisher.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2018 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation; either version 2 of the License, or (at your option) any later
+# version.
+#
+# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+# Suite 330, Boston, MA 02111-1307, USA.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+
+import json
+import logging
+
+import pika
+
+from reana_workflow_engine_serial.config import (BROKER_PASS, BROKER_PORT,
+                                                 BROKER_URL, BROKER_USER,
+                                                 EXCHANGE, ROUTING_KEY,
+                                                 STATUS_QUEUE)
+
+
+class Publisher:
+    """Progress publisher to MQ."""
+
+    def __init__(self):
+        self.broker_credentials = pika.PlainCredentials(BROKER_USER,
+                                                        BROKER_PASS)
+        self._params = pika.connection.ConnectionParameters(
+            BROKER_URL, BROKER_PORT, '/', self.broker_credentials)
+        self._conn = None
+        self._channel = None
+
+    def connect(self):
+        """Connect to MQ channel."""
+        if not self._conn or self._conn.is_closed:
+            self._conn = pika.BlockingConnection(self._params)
+            self._channel = self._conn.channel()
+            self._channel.queue_declare(queue=STATUS_QUEUE)
+
+    def _publish(self, msg):
+        self._channel.basic_publish(exchange=EXCHANGE,
+                                    routing_key=ROUTING_KEY,
+                                    body=json.dumps(msg))
+        logging.debug('Publisher: message sent: %s', msg)
+
+    def publish_workflow_status(self, workflow_uuid, status,
+                                logs='', message=None):
+        """Publish msg, reconnecting if necessary."""
+        try:
+            msg = {"workflow_uuid": workflow_uuid,
+                   "logs": logs,
+                   "status": status,
+                   "message": message}
+            self._publish(msg)
+        except pika.exceptions.ConnectionClosed:
+            logging.debug('Publisher: ConnectionClosed, reconnecting.')
+            print('Publisher: ConnectionClosed, reconnecting.')
+            self.connect()
```

Comment:
Is it enough to try to reconnect only twice?

Reply:
FWIW it worked on my cluster. 

```console
$ kubectl get pods
NAME                                     READY     STATUS    RESTARTS   AGE
cwl-default-worker-8494bdcfc4-ckpg8      1/1       Running   2          7m
db-54bd479dd8-pbndf                      1/1       Running   0          7m
job-controller-65f87df9df-kzmcs          1/1       Running   0          8m
message-broker-69ff599898-qqcl9          1/1       Running   0          8m
serial-default-worker-5788fb67f5-v96fc   1/1       Running   0          7m
server-645485b68-tprtg                   1/1       Running   2          8m
wdb-8b76ff44b-f8gzn                      1/1       Running   0          7m
workflow-controller-76d7b87887-7kblz     2/2       Running   2          8m
workflow-monitor-84bcfd65c7-wvjt8        1/1       Running   0          8m
yadage-default-worker-7d5dbf956d-mls49   1/1       Running   0          7m
zeromq-msg-proxy-6c4f6bd6b-2h4lc         1/1       Running   0          7m
```

The fact that the server pod restarts may be a bit annoying, since it means that the endpoint may change and that people have to do `eval $(reana-cluster env)` twice...

Reply:
@diegodelemos good question, the reconnection attempt will happen for each outgoing message if necessary though, so the engine shouldn't get completely disconnected. Still, it is possible that if a specific update gets lost the count is wrong from that point on. Also, by adding a `while not connected, retry` loop, we risk getting stuck workflows, because of issues in the message broker.